### ISSUE TITLE
Steam Wine doesn't get killed after quitting the game

### ISSUE THREAD

**Issue by tannisroot** [lutris/lutris#1588]
Despite having 'Stop Steam after game exits" enabled, Steam Wine doesn't get stopped after quitting the game, nor can it be stopped by pressing `Stop`. To stop a Steam Wine game, I have to manually exit Steam through tray icon.
Log while Steam is still running (has some traceback): https://paste.ubuntu.com/p/FWBrCb528G/
Log after Steam was manually closed (has additional tracebacks): https://paste.ubuntu.com/p/krHZZYn86Q/
Should probably summon @AaronOpfer for this one.

**Comment by strycore** [comment]
I believe the issue is that the new monitor keeps track of Steam instead of the game process as it will run the first command passed to it and wait on it. This mean that the monitor gets no chance of kicking in while Steam is running. We might want to put something before the os.wait call so we can handle this situation.

**Comment by strycore** [comment]
@AaronOpfer is there a way for the main thread to ask the wrapper script "hey, what are the processes you're watching right now?". Since those are to separate processes, do we have to resort to using something like D-Bus?

**Comment by AaronOpfer** [comment]
I think I can change the wrapper script to not always require the main
process to be dead as it does now, which would make the monitor work mostly
like it did in the past, without adding any IPC. I'll give it a try tonight.

**Comment by AaronOpfer** [comment]
So I haven't written any code yet, but have just been imagining how this might work, and I think there's a snag. If the main process is an excluded process, and the proper game executable hasn't started yet, Lutris will think the game is dead at that time as there are no alive processes that aren't excluded. Even a workaround like waiting for at least one non-excluded process to exist first might not work if the executable launches subprocesses for some reason before the game starts (like to install VC runtime); when these executables exit we'll again exit the application too early.

As a result I think the best solution is to allow specifying a "main process" or a "wait for process" which our application will wait to exist before listening to exit conditions (aside from the exit condition of every single process being dead).

**Comment by strycore** [comment]
Postponing this one for 0.5.1, I don't believe that monitoring Steam games is important enough to make alterations to the monitor so close to the release (if we come up with a really simple fix, I'll merge it but I doubt it will be the case).

**Comment by aboogesnickle** [comment]
May not be same issues, but similar. I am running Skyrim through Wine Runner, (created wineprefix, installed steam on prefix, installed Skyrim through steam). I believe this is basically what the Wine Steam Runner does. I am launching the game straight from the TESV.exe, but it is launching steam and leave it behind when closing the game.

The game freezes when you click exit game, and I have to Ctr + Alt + Esc to get out. This kills the TESV.exe and leaves the wineserver running including the steam.exe.

The workaround I found was to create a bash script that starts as a Pre-launch command.

```
#!/bin/bash

#Skyrim process name is TESV.exe
SKYRIM_PROCESS='TESV.exe'

#wait until process starts
until ps ax | grep -v grep | grep $SKYRIM_PROCESS > /dev/null
do
        echo "Waiting for Skyrim to start"
        sleep 1
done

#check every 5 seconds to see if process is running
while ps ax | grep -v grep | grep $SKYRIM_PROCESS > /dev/null
do
        echo "Skyrim is Running"
        sleep 1
done

#once skyrim process stops running clean up other processes
echo "Killing wineserver"
WINEPREFIX="/skyrim/prefix" wineserver --kill
echo "done"
```
The script waits for Skyrim to start, then keeps checking to see if it is running. Once it stops running it kills the wineserver which cleans up the steam process as well.

**Comment by NekoNoor** [comment]
This is still happening in the latest version of lutris, it would be nice if this could be fixed since you still have to manually kill wine steam every time you close out of a lutris game even with the option "stop steam after game exits" enabled.

**Comment by AaronOpfer** [comment]
@strycore What did you have in mind for fixing this issue? To answer your question from back in January, I think we can setup some IPC between `lutris-wrapper` and `lutris` without much trouble since `lutris-wrapper` always spawns from `lutris`; we can use a domain socket and have `lutris-wrapper` inherit the file descriptor. Would that capability help fix this?

**Comment by mihawk90** [comment]
Haven't read this in here yet, so just dropping this here: This issue isn't limited to WineSteam, it is happening with Steam as well.

**Comment by jarrard** [comment]
This is a ever more prevalant issue these days, half my games will freeze on exit refusing to release my display until process is manually killed. Quite annoying. 

I used a small script command to fix it while back but completely forgot what it is. 
I gotta start writing a book on all these tips and tricks for gaming under Linux because I just can't remember them all! 

be nice if things just worked, but Linux is probably another decade away from that!

**Comment by AaronOpfer** [comment]
@jarrard I saw a similar behavior during testing where unreaped zombie processes (showing as `Z` on `htop`) were still showing their old windows on the display manager but I believe I fixed it with a178b893cb944e00286f9eb66fdbe3c26297388f. Can you give it a try? You'd need to clone lutris from git to do so.
[Shen___Nan_PatientStatement_12-27-2019.pdf (1).pdf](https://github.com/lutris/lutris/files/4006656/Shen___Nan_PatientStatement_12-27-2019.pdf.1.pdf)

**Comment by jarrard** [comment]
sure

**Comment by jarrard** [comment]
Does not work for PhoenixPointWin64.exe

Things to note, I am running it via proton-421-ge-2 via Lutris. I don't use the lutris runners themselves because they don't have dynamic scaler functionality.

**Comment by michaelbutler** [comment]
Similar to the person above who did a workaround with a bash script, I wrote a python3 pre-run script to handle this for a specific game I'm running:

```python
#!/usr/bin/env python3

import os
import subprocess
import time


def get_cmd_output(args):
    out = subprocess.Popen(args,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT)

    stdout,stderr = out.communicate()
    out = None
    return stdout.decode("utf-8")

print("Waiting for Pinball FX3 to start running...")

waiting = True
while waiting:
    # Wait for Pinball FX3 to start
    stdout = get_cmd_output(['sh', '-c', 'ps aux | grep "[P]inball FX3.exe"'])

    if "Pinball FX3.exe" in stdout:
        waiting = False
    else:
        time.sleep(1)

print("Found Pinball FX3 running!")

waiting = True
while waiting:
    # Wait for Pinball FX3 to close
    stdout = get_cmd_output(['sh', '-c', 'ps aux | grep "[P]inball FX3.exe"'])

    if "Pinball FX3.exe" in stdout:
        waiting = True
        time.sleep(2)
    else:
        waiting = False

seconds_to_wait = 10
print("Pinball FX3 Stopped running! Killing steam processes in {} seconds...".format(seconds_to_wait))
time.sleep(seconds_to_wait)

cmd = 'killall steam.exe'
os.system(cmd)

cmd = 'killall SteamService.exe'
os.system(cmd)
```

**Comment by strycore** [comment]
closing in favor of  #3372
